client/utils/src/notification.rs

```diff
@@ -44,108 +48,93 @@ pub trait TracingKeyStr {
-/// used to add more subscriptions.
-#[derive(Clone)]
-pub struct NotificationStream<Payload: Clone, TK: TracingKeyStr> {
-	subscribers: SharedSenders<Payload>,
-	_trace_key: PhantomData<TK>,
+/// Used to send notifications from the BEEFY gadget side.
+pub struct NotificationSender<Payload> {
+	hub: Hub<TracingUnbounded<Payload>, Registry>,
 }
 
-impl<Payload: Clone, TK: TracingKeyStr> NotificationStream<Payload, TK> {
+impl<Payload, TK: TracingKeyStr> NotificationStream<Payload, TK> {
 	/// Creates a new pair of receiver and sender of `Payload` notifications.
 	pub fn channel() -> (NotificationSender<Payload>, Self) {
-		let subscribers = Arc::new(Mutex::new(vec![]));
-		let receiver = NotificationStream::new(subscribers.clone());
-		let sender = NotificationSender::new(subscribers);
+		let channels = TracingUnbounded::new(TK::TRACING_KEY);
+		let hub = Hub::new(channels);
+		let sender = NotificationSender { hub: hub.clone() };
+		let receiver = NotificationStream { hub, _pd: Default::default() };
 		(sender, receiver)
 	}
 
-	/// Create a new receiver of `Payload` notifications.
-	///
-	/// The `subscribers` should be shared with a corresponding `NotificationSender`.
-	fn new(subscribers: SharedSenders<Payload>) -> Self {
-		Self { subscribers, _trace_key: PhantomData }
-	}
-
 	/// Subscribe to a channel through which the generic payload can be received.
-	pub fn subscribe(&self) -> TracingUnboundedReceiver<Payload> {
-		let (sender, receiver) = tracing_unbounded(TK::TRACING_KEY);
-		self.subscribers.lock().push(sender);
-		receiver
+	pub fn subscribe(&self) -> NotificationReceiver<Payload> {
+		let receiver = self.hub.subscribe(());
+		NotificationReceiver { receiver }
 	}
 }
 
-#[cfg(test)]
-mod tests {
-	use super::*;
-	use futures::StreamExt;
-
-	#[derive(Clone)]
-	pub struct DummyTracingKey;
-	impl TracingKeyStr for DummyTracingKey {
-		const TRACING_KEY: &'static str = "test_notification_stream";
-	}
+impl<Payload> NotificationSender<Payload> {
+	/// Send out a notification to all subscribers that a new payload is available for a
+	/// block.
+	pub fn notify<Error>(
+		&self,
+		payload: impl FnOnce() -> Result<Payload, Error>,
+	) -> Result<(), Error>
+	where
+		Payload: Clone,
+	{
+		// The subscribers collection used to be cleaned upon send previously.
+		// The set used to be cleaned up twice:
+		// - once before sending: filter on `!tx.is_closed()`;
+		// - once while sending: filter on `!tx.unbounded_send().is_err()`.
+		//
+		// Since there is no `close` or `disconnect` operation defined on the
+		// `NotificationReceiver<Payload>`, the only way to close the `rx` is to drop it.
+		// Upon being dropped the `NotificationReceiver<Payload>` unregisters its `rx`
+		// from the registry using its `_subs_guard`.
+		//
+		// So there's no need to clean up the subscribers set upon sending another message.
+
+		let payload = payload()?; // FIXME: Did it have to be lazily instantiated?
```

Comment:
If you read the old code, you see that this was done to not construct the payload when there would be no subscriber. I'm not sure about the costs of the payload, but we should keep it the way.

Reply:
Bringing lazy payload materialisation back can be done quite simply: https://github.com/RGafiyatullin/substrate/commit/f51fecc1656f5ffe853ca15795095187979e2102 .


But, in my opinion, providing such API can lead to a surprise when the user isn't duly alert.

So I peeked into the current usages.

Some of the notifications are hashes, I believe those are not tremendously expensive to clone in advance (even if no one is listening):

- https://github.com/paritytech/substrate/blob/master/client/beefy/src/worker.rs#L252
- https://github.com/paritytech/substrate/blob/master/client/beefy/src/worker.rs#L377
- https://github.com/paritytech/substrate/blob/master/client/beefy/rpc/src/lib.rs#L260 (in a test)

Some are structures having a `Vec` field, those could benefit more if they were `Arc`-ed:

- https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/environment.rs#L1334 
- https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/rpc/src/lib.rs#L472 (in a test)
- https://github.com/paritytech/substrate/blob/master/client/beefy/src/worker.rs#L370
- https://github.com/paritytech/substrate/blob/master/client/beefy/rpc/src/lib.rs#L375 (in a test)

Reply:
Okay, then at least remove the Fixme comment.

Reply:
For now, I will keep the API as it was before I touched it.

Altering the notification message-types to be `Arc`-ed for the potentially heavy objects would be too much of an invasion for now.